src/services/ssdi.ts

```diff
@@ -1,197 +1,697 @@
-  {
-    bind: [`ssdi.${key}`, JSON.stringify(value)],
-    type: db.sequelize.QueryTypes.SELECT,
-  },
-)));
-
-// Helper function to sanitize filenames
-const sanitizeFilename = (filename: string): string => filename.replace(/[^a-zA-Z0-9-_]/g, '_');
-
-// Helper function to generate a string representation of the flag values
-const generateFlagString = (flagValues: FlagValues): string => Object.entries(flagValues)
-  .map(([key, value]) => `${key}_${Array.isArray(value) ? value.join('-') : value}`)
+// Helper function to set filters in the database
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const setFilters = async (filterValues: FilterValues): Promise<any[]> => Promise.all(
+  Object.entries(filterValues).map(async ([key, value]) => {
+    if (key.endsWith('.not')) {
+      const baseKey = key.slice(0, -4);
+
+      return Promise.all([
+        db.sequelize.query(
+          'SELECT set_config($1, $2, false)',
+          {
+            bind: [`ssdi.${baseKey}`, JSON.stringify(value)],
+            type: db.sequelize.QueryTypes.SELECT,
+          },
+        ),
+        db.sequelize.query(
+          'SELECT set_config($1, $2, false)',
+          {
+            bind: [`ssdi.${key}`, JSON.stringify(true)],
+            type: db.sequelize.QueryTypes.SELECT,
+          },
+        ),
+      ]);
+    }
+    return db.sequelize.query(
+      'SELECT set_config($1, $2, false)',
+      {
+        bind: [`ssdi.${key}`, JSON.stringify(value)],
+        type: db.sequelize.QueryTypes.SELECT,
+      },
+    );
+  }),
+);
+
+// Helper function to generate a string representation of the filter values
+const generateFilterString = (filterValues: FilterValues): string => Object.entries(filterValues)
+  .map(([key, value]) => {
+    const formattedValue = Array.isArray(value) ? value.join('-') : value;
+    return `${key}_${formattedValue}`;
+  })
   .join('_');
 
+// Execute query asynchronously and set read-only transaction
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-const executeQuery = async (query: string): Promise<any> => {
+const executeQuery = async (filePath: string): Promise<any> => {
+  const resolvedFilePath = safeResolvePath(filePath);
+
+  if (!cache.has(resolvedFilePath)) {
+    const cachedFile = await readJsonHeaderFromFile(resolvedFilePath);
+    if (cachedFile) {
+      cache.set(resolvedFilePath, cachedFile);
+    } else {
+      throw new Error(`Unable to read and parse the JSON header from file: ${resolvedFilePath}`);
+    }
+  }
+
+  const cachedFile = cache.get(resolvedFilePath) as CachedFile;
+  const { query } = cachedFile;
+
   if (typeof query !== 'string') {
     throw new Error('The query must be a string');
   }
 
   try {
-    // Set transaction to READ ONLY, this will fail the transaction if any tables are modified
-    await db.sequelize.query('SET TRANSACTION READ ONLY;', { type: QueryTypes.RAW });
-
     const result = await db.sequelize.query(query, { type: QueryTypes.SELECT });
```

Comment:
other dashboard queries are cached in redis, we may want to consider this (as it helps protect against issues arising from concurrent usage)

Reply:
Good point, I will look at how to add that